### PR TITLE
raft: send follower's commit index in MsgAppResp

### PR DESCRIPTION
**Pull request by iskettaneh**
This commit makes the follower send the commit index in the MsgAppResp. This will be used in future commits to make the leader determine if a follower's commit index is up-to-date or not, and decide whether it needs an empty MsgApp based on that.

References: #125266

Epic: None

Release note: None

### CODE REVIEW

File: pkg/raft/raft.go [hunk 1]
```diff
@@ -1722,8 +1724,10 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 		// committed entries at m.Term (by raft invariants), so it is safe to bump
 		// the commit index even if the MsgApp is stale.
 		lastIndex := a.lastIndex()
-		r.raftLog.commitTo(logMark{term: m.Term, index: min(m.Commit, lastIndex)})
-		r.send(pb.Message{To: m.From, Type: pb.MsgAppResp, Index: lastIndex})
+		commitIndex := min(m.Commit, lastIndex)
+		r.raftLog.commitTo(logMark{term: m.Term, index: commitIndex})
```
Comment (pav-kv) [inline pkg/raft/raft.go:1728]: nit: the `commitTo` line did not need to change (`commitIndex` doesn't have a second usage)

File: pkg/raft/raft.go [hunk 2]
```diff
@@ -1751,6 +1755,7 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 		To:         m.From,
 		Type:       pb.MsgAppResp,
 		Index:      m.Index,
+		Commit:     r.raftLog.committed,
```
Comment (pav-kv) [inline pkg/raft/raft.go:1758]: Do we need to attach `Commit` to reject resps? I think it's fine/consistent, but just want to see what others think.

One way to think about the commit index flow is that it's independent from log appends. So it doesn't matter if the append is accepted/rejected, we just use its flow as a side-channel.

**Review by pav-kv — APPROVED** [review]

**Comment by pav-kv** [comment]
With this PR, the communicated `Commit` index is not guaranteed to be synced yet, correct?

**Review by iskettaneh — COMMENTED** [review]
Correct. A follow-up PR should do that. However, this should be safe since we don't use this Commit index yet.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-:-O3mAA7r5vJdpmV8uZdj:b-eez0z7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @pav-kv)

___
*[`pkg/raft/raft.go` line 1728 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128497#gh-1709149544:-O3mDFrD-h6lWCUFaNHR:b3f7uwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66d1a9b9a46ccb076c81be6e4421fb20eae73c5/pkg/raft/raft.go#L1728)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: the `commitTo` line did not need to change (`commitIndex` doesn't have a second usage)
</blockquote></details>

Yeah my bad, I was experimenting with something and forgot to revert it.
___
*[`pkg/raft/raft.go` line 1758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128497#gh-1709203256:-O3mAOTHBydo3uqN5aSo:b-sdvrix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66d1a9b9a46ccb076c81be6e4421fb20eae73c5/pkg/raft/raft.go#L1758)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Do we need to attach `Commit` to reject resps? I think it's fine/consistent, but just want to see what others think.

One way to think about the commit index flow is that it's independent from log appends. So it doesn't matter if the append is accepted/rejected, we just use its flow as a side-channel.
</blockquote></details>

That's a good question. I also interested to see what the others think. But I agree with you that we can remove it from the rejected case and still be okay.

**Review by nvb — COMMENTED** [review]
> Reviewed 24 of 25 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @pav-kv)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-O3mS3hf7vkN4OlG-SQK):*
When making this change, we'll also need to update the commentary in `Replica.sendRaftMessages`, which says that "A successful (non-reject) MsgAppResp contains one piece of information ...". The logic there should still be ok.
___
*[`pkg/raft/raft.go` line 1758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128497#gh-1709203256:-O3mRD0n7ZA80K8eCdMS:b-swn6rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66d1a9b9a46ccb076c81be6e4421fb20eae73c5/pkg/raft/raft.go#L1758)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

That's a good question. I also interested to see what the others think. But I agree with you that we can remove it from the rejected case and still be okay.
</blockquote></details>

Agreed about commit index flow being independent from log appends. That means that the commit index should be allowed on rejected messages. It also means that the commit index should be optional — if a follower doesn't provide it then a leader just can't advance it's tracking.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-O3mS3hf7vkN4OlG-SQK):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

When making this change, we'll also need to update the commentary in `Replica.sendRaftMessages`, which says that "A successful (non-reject) MsgAppResp contains one piece of information ...". The logic there should still be ok.
</blockquote></details>

[Out of scope for this PR] The logic there is potentially suboptimal. It says that a higher index response wins, but it should instead be the highest `(term, index)`. It's possible that a higher-term leader overwritten a part of the log and the last index regressed.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-O3mS3hf7vkN4OlG-SQK):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

[Out of scope for this PR] The logic there is potentially suboptimal. It says that a higher index response wins, but it should instead be the highest `(term, index)`. It's possible that a higher-term leader overwritten a part of the log and the last index regressed.
</blockquote></details>

https://github.com/cockroachdb/cockroach/blob/a976e112c91b57182921dfc51f96aaec34ec8250/pkg/kv/kvserver/replica_raft.go#L1721-L1724

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-O3mS3hf7vkN4OlG-SQK):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/blob/a976e112c91b57182921dfc51f96aaec34ec8250/pkg/kv/kvserver/replica_raft.go#L1721-L1724
</blockquote></details>

Created #128604 for this, let's discuss there.

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @pav-kv)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-O3mS3hf7vkN4OlG-SQK):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Created #128604 for this, let's discuss there.
</blockquote></details>

Change the comment. Will keep an eye on #128604 to check if the logic needs to be changed.
___
*[`pkg/raft/raft.go` line 1758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128497#gh-1709203256:-O3n8HOV5sqOWUWT9Jsa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66d1a9b9a46ccb076c81be6e4421fb20eae73c5/pkg/raft/raft.go#L1758)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Agreed about commit index flow being independent from log appends. That means that the commit index should be allowed on rejected messages. It also means that the commit index should be optional — if a follower doesn't provide it then a leader just can't advance it's tracking.
</blockquote></details>

Done.

**Review by pav-kv — APPROVED** [review]
Should we also include `Commit` into [this](https://github.com/cockroachdb/cockroach/blob/deb8ec102da0867e50d51e378010b98ff208b95b/pkg/raft/raft.go#L1127) message that you referenced in another PR? More generally, should we just include it in all occurrences of `MsgAppResp` (if not already)?

Btw, one alternative way to achieve this (instead of adding the `Commit` into all occurrences of `MsgAppResp` in this PR) is to do it in the `send` [interceptor](https://github.com/cockroachdb/cockroach/blob/e103ae12501b6d4f5e4ffd235bc4a3ecc5b20672/pkg/raft/raft.go#L488-L569). It's indirect though, not sure if it's better. Up to you.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-:-O3nTH3e0ZzQyOlAqEIK:b6f3d3b)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

**Review by iskettaneh — COMMENTED** [review]
I didn't populate the commit index in that case on purpose. Since that case is a very special case where when the leader receive it, it will step down (it has a higher term compared to the leader).
That said, I don't mind always populating the commit index (similar to how we always populate the term)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-:-O3rJqdW4Z6ssrHqaVrF:bmq4b9w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @pav-kv)

**Review by pav-kv — APPROVED** [review]
Either way is fine. It's important to include the commit index into the messages that guarantee liveness/convergence of the commit index, that is same-term non-reject `MsgAppResp`s. Other cases are optional, and probably don't give much advantage except in some rare cases.

For example, this special case that you mentioned:

> very special case where when the leader receive it, it will step down (it has a higher term compared to the leader)

It may happen that this message was stale, and the receiver of this `MsgAppResp` message is actually already a leader of this higher term. Then this message would bring some useful information to this leader. But it will get this information otherwise anyway, so it's not critical to over-optimize for this case.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-:-O3rzdQT2m3GE-rkeqgT:b-bup31z)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

**Review by arulajmani — APPROVED** [review]
> Reviewed 2 of 12 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @pav-kv)

___
*[`pkg/raft/raft.go` line 1758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128497#gh-1709203256:-O47j8Ug9OunWZFU-yjT:b-e5b2rw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66d1a9b9a46ccb076c81be6e4421fb20eae73c5/pkg/raft/raft.go#L1758)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Consider adding a short comment right above this field assignment calling this out and why.

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @pav-kv)

___
*[`pkg/raft/raft.go` line 1758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/128497#gh-1709203256:-O4B5jJzDYsEBSvVvXnQ:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66d1a9b9a46ccb076c81be6e4421fb20eae73c5/pkg/raft/raft.go#L1758)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding a short comment right above this field assignment calling this out and why.
</blockquote></details>

Done!

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 11 of 12 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128497#-:-O4GEJRy-1nRlVjfAwk8:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv)

**Pull request merged**